src/core/src/op/unsqueeze.cpp

```diff
@@ -33,120 +31,61 @@ void op::v0::Unsqueeze::validate_and_infer_types() {
-// evaluate and inference can use same function to calculate output shape. TODO for next version for this operator.
-namespace {
-bool evaluate_unsqueeze(const Node* node,
-                        const HostTensorPtr& arg0,
-                        const HostTensorPtr& arg1,
-                        const HostTensorPtr& out) {
-    auto element_type = arg0->get_element_type();
-    out->set_element_type(element_type);
-
-    const auto& axes_shape = arg1->get_shape();
-    ov::op::v0::check_unsqueeze_axes_rank(node, Rank(axes_shape.size()));
-
-    const auto& data_shape = arg0->get_shape();
-    const auto out_rank = static_cast<int64_t>(data_shape.size() + shape_size(axes_shape));
-
-    // Get axes and normalize
-    OPENVINO_SUPPRESS_DEPRECATED_START
-    auto axes = read_index_vector(arg1);
-    normalize_axes(node, out_rank, axes);
-    OPENVINO_SUPPRESS_DEPRECATED_END
-
-    // Sort in increasing order
-    std::set<int64_t> axes_set(axes.begin(), axes.end());
-    NGRAPH_CHECK(axes.size() == axes_set.size(), "Axes has duplicate axis.");
-
-    auto out_shape = data_shape;
-    for (int64_t axis : axes_set) {
-        out_shape.insert(out_shape.begin() + axis, 1);
-    }
-    out->set_shape(out_shape);
-
-    using namespace ov::element;
-    return IfTypeOf<i32, i64, u32, u64, f16, f32, f64, bf16>::apply<Evaluate>(element_type,
-                                                                              arg0,
-                                                                              out,
-                                                                              shape_size(out_shape));
+    return std::make_shared<Unsqueeze>(new_args.at(0), new_args.at(1));
 }
-}  // namespace
-}  // namespace unsqueeze
-}  // namespace op
-}  // namespace ov
 
-bool op::v0::Unsqueeze::evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs) const {
+bool ov::op::v0::Unsqueeze::evaluate(ov::TensorVector& outputs, const ov::TensorVector& inputs) const {
     OV_OP_SCOPE(v0_Unsqueeze_evaluate);
-    OPENVINO_SUPPRESS_DEPRECATED_START
-    NGRAPH_CHECK(validate_host_tensor_vector(inputs, 2));
-    NGRAPH_CHECK(validate_host_tensor_vector(outputs, 1));
-    OPENVINO_SUPPRESS_DEPRECATED_END
-    return unsqueeze::evaluate_unsqueeze(this, inputs[0], inputs[1], outputs[0]);
+    OPENVINO_ASSERT(inputs.size() == 2);
+    if (outputs.empty()) {
+        outputs.emplace_back(ov::Tensor(inputs[0].get_element_type(), {0}));
+    } else {
+        OPENVINO_ASSERT(outputs.size() == 1);
+    }
+    const auto& output_shape = shape_infer(this,
+                                           std::vector<ov::PartialShape>{inputs[0].get_shape(), inputs[1].get_shape()},
+                                           make_tensor_accessor(inputs))
+                                   .front()
+                                   .to_shape();
+    outputs[0].set_shape(output_shape);
+    ov::reference::copy(static_cast<const char*>(inputs[0].data()),
```

Comment:
@ilyachur, Just need opinion:
If reference is not working on specific type maybe it should use `void*` instead `char*`, then casting from tensor will be not required. It will be also more compatible with `std::memcpy`